#Álgebra lineal usando intervalos

###Autor: Leonardo Andrés Castro Enríquez

In [1]:
using Intervalos

In [2]:
using FactCheck

In [3]:
include("MatricesTest.jl")

Prueba: Álgebra lineal con intervalos
12 facts verified.


delayed_handler (generic function with 4 methods)

####Definimos las matrices A y B de dimensión 2x2 cuyos elementos son intervalos

In [5]:
A=[Intervalo(0,0.5) Intervalo(1,1.5); Intervalo(2,2.5) Intervalo(3,3.5)]

2x2 Array{Intervalo,2}:
 (0e+00, 5e-01)    (1e+00, 1.5e+00)
 (2e+00, 2.5e+00)  (3e+00, 3.5e+00)

In [6]:
B=[Intervalo(4,4.5) Intervalo(5,5.5); Intervalo(6,6.5) Intervalo(7,7.5)]

2x2 Array{Intervalo,2}:
 (4e+00, 4.5e+00)  (5e+00, 5.5e+00)
 (6e+00, 6.5e+00)  (7e+00, 7.5e+00)

##Operaciones entre matrices

###Suma

In [27]:
C=A+B

2x2 Array{Intervalo,2}:
 (4e+00, 5e+00)  (6e+00, 7e+00)  
 (8e+00, 9e+00)  (1e+01, 1.1e+01)

In [19]:
B+A

2x2 Array{Intervalo,2}:
 (4e+00, 5e+00)  (6e+00, 7e+00)  
 (8e+00, 9e+00)  (1e+01, 1.1e+01)

In [28]:
#cumple la propiedad de conmutatividad
A+B==B+A

true

###Multiplicación 

In [8]:
A*B

2x2 Array{Intervalo,2}:
 (6e+00, 1.2e+01)    (7e+00, 1.4e+01)
 (2.6e+01, 3.4e+01)  (3.1e+01, 4e+01)

In [18]:
B*A

2x2 Array{Intervalo,2}:
 (1e+01, 1.6e+01)    (1.9e+01, 2.6e+01)
 (1.4e+01, 2.2e+01)  (2.7e+01, 3.6e+01)

In [30]:
#Cumple la propiedad de anticonmutatividad 
A*B!=B*A

true

In [9]:
A

2x2 Array{Intervalo,2}:
 (0e+00, 5e-01)    (1e+00, 1.5e+00)
 (2e+00, 2.5e+00)  (3e+00, 3.5e+00)

In [10]:
B

2x2 Array{Intervalo,2}:
 (4e+00, 4.5e+00)  (5e+00, 5.5e+00)
 (6e+00, 6.5e+00)  (7e+00, 7.5e+00)

In [9]:
#Multiplicamos la primer componente de A por la primer componente de B
a=Intervalo(0,0.5)*Intervalo(4,4.5)

(0e+00, 2.25e+00)

In [10]:
#Multiplicamos la segunda componente de A por la segunda de B
b=Intervalo(1,1.5)*Intervalo(6,6.5)

(6e+00, 9.75e+00)

In [11]:
#Cabe resaltar que la multiplicación entre intervalos tiene su propia álgebra (definida en el módulo Intervalos). 
#Finalmente, sumamos los resultados y obtenemos la primer componente de A*B
a+b

(6e+00, 1.2e+01)

####otro ejemplo [1]: 

In [11]:
C=[Intervalo(1,2) Intervalo(3,4)]

1x2 Array{Intervalo,2}:
 (1e+00, 2e+00)  (3e+00, 4e+00)

In [12]:
D=[Intervalo(5,6) Intervalo(7,8); Intervalo(9,10) Intervalo(11,12)]

2x2 Array{Intervalo,2}:
 (5e+00, 6e+00)  (7e+00, 8e+00)    
 (9e+00, 1e+01)  (1.1e+01, 1.2e+01)

In [13]:
C*D

1x2 Array{Intervalo,2}:
 (3.2e+01, 5.2e+01)  (4e+01, 6.4e+01)

####[1] R. E. Moore,  R. B. Kearfott, M. J. Cloud, "Introduction to Interval Analysis",  SIAM editorial, Philadelphia, Cap.7, Secc.7.2, pág. 86, 2009.

###Multiplicación por un entero

In [12]:
A*2

2x2 Array{Intervalo,2}:
 (0e+00, 1e+00)  (2e+00, 3e+00)
 (4e+00, 5e+00)  (6e+00, 7e+00)

###Traza

In [13]:
A

2x2 Array{Intervalo,2}:
 (0e+00, 5e-01)    (1e+00, 1.5e+00)
 (2e+00, 2.5e+00)  (3e+00, 3.5e+00)

In [14]:
Tr(A)

(3e+00, 4e+00)

###Transpuesta

In [15]:
A

2x2 Array{Intervalo,2}:
 (0e+00, 5e-01)    (1e+00, 1.5e+00)
 (2e+00, 2.5e+00)  (3e+00, 3.5e+00)

In [32]:
T(A)

2x2 Array{Intervalo,2}:
 (0e+00, 5e-01)    (2e+00, 2.5e+00)
 (1e+00, 1.5e+00)  (3e+00, 3.5e+00)

###Propiedades de la transpuesta 

$(A^{T})^{T} = A$

In [31]:
T(T(A))==A

true

$(A+B)^{T} = A^{T} + B^{T}$

In [35]:
T(A+B)==T(A)+T(B)

true

$(\lambda A)^{T}=\lambda A^{T}$ con $\lambda$ un real

In [37]:
T(A*2)==T(A)*2

true

$(AB)^{T}=B^{T}A^{T}$

In [38]:
T(A*B)==T(B)*T(A)

true

$(A^{-1})^{T}=(A^{T})^{-1}$

In [40]:
T(inv(A))==inv(T(A))

true

###Determinante

In [9]:
A

2x2 Array{Intervalo,2}:
 (0e+00, 5e-01)    (1e+00, 1.5e+00)
 (2e+00, 2.5e+00)  (3e+00, 3.5e+00)

In [10]:
Det(A)

(-2e+00, -2e+00)

In [19]:
#multiplicación de los intervalos en las posiciones A[1] y A[4] 
c=Intervalo(0,0.5)*Intervalo(3,3.5)

(0e+00, 1.75e+00)

In [20]:
#multiplicación de los intervalos en las posiciones A[3] y A[2]
d=Intervalo(1,1.5)*Intervalo(2,2.5)

(2e+00, 3.75e+00)

In [21]:
#Finalmente, restamos los resultados obtenidos anteriormente
c-d

(-2e+00, -2e+00)

###Adjunta

In [22]:
A

2x2 Array{Intervalo,2}:
 (0e+00, 5e-01)    (1e+00, 1.5e+00)
 (2e+00, 2.5e+00)  (3e+00, 3.5e+00)

In [23]:
adj(A)

2x2 Array{Intervalo,2}:
 (3e+00, 3.5e+00)    (-2.5e+00, -2e+00)
 (-1.5e+00, -1e+00)  (0e+00, 5e-01)    

###Inversa

In [39]:
inv(A)

2x2 Array{Intervalo,2}:
 (-1.75e+00, -1.5e+00)  (5e-01, 7.5e-01)  
 (1e+00, 1.25e+00)      (-2.5e-01, -0e+00)

##Operador densidad

El operador densidad se define en términos del ket $|\psi\rangle$ y bra $\langle \psi|$ como

\begin{equation}
\rho = |\psi\rangle \langle \psi| = \displaystyle\binom{a}{b} \big(a^* b^* \big)= \begin{pmatrix} aa^* & ab^* \\ ba^* & bb^* \end{pmatrix}. 
\end{equation} 

Definiendo el ket y el bra en términos de intervalos, tenemos

In [12]:
[Intervalo(1,2); Intervalo(3,4)] 

2-element Array{Intervalo,1}:
 (1e+00, 2e+00)
 (3e+00, 4e+00)

In [7]:
#normalizando
Intervalo(1,2)^2+Intervalo(3,4)^2

(1e+01, 2e+01)

In [8]:
(ans)^1/2

(5e+00, 1e+01)

In [13]:
#Por lo tanto el ket normalizado es de la forma
psi1=[Intervalo(1,2)/Intervalo(5,10); Intervalo(3,4)/Intervalo(5,10)]

2-element Array{Intervalo,1}:
 (9.999999999999999999999999999999999999999999999999999999999999999999999999999914e-02, 4.000000000000000000000000000000000000000000000000000000000000000000000000000009e-01)
 (2.999999999999999999999999999999999999999999999999999999999999999999999999999974e-01, 8.000000000000000000000000000000000000000000000000000000000000000000000000000017e-01)

In [14]:
#Por otro lado el bra queda
psi2=transpose(psi1)

1x2 Array{Intervalo,2}:
 (9.999999999999999999999999999999999999999999999999999999999999999999999999999914e-02, 4.000000000000000000000000000000000000000000000000000000000000000000000000000009e-01)  …  (2.999999999999999999999999999999999999999999999999999999999999999999999999999974e-01, 8.000000000000000000000000000000000000000000000000000000000000000000000000000017e-01)

In [15]:
#Por consiguente, obtenemos la matriz densidad
rho=psi1*psi2

2x2 Array{Intervalo,2}:
 (9.999999999999999999999999999999999999999999999999999999999999999999999999999725e-03, 1.600000000000000000000000000000000000000000000000000000000000000000000000000021e-01)  …  (2.999999999999999999999999999999999999999999999999999999999999999999999999999931e-02, 3.200000000000000000000000000000000000000000000000000000000000000000000000000041e-01)
 (2.999999999999999999999999999999999999999999999999999999999999999999999999999931e-02, 3.200000000000000000000000000000000000000000000000000000000000000000000000000041e-01)     (8.999999999999999999999999999999999999999999999999999999999999999999999999999793e-02, 6.400000000000000000000000000000000000000000000000000000000000000000000000000083e-01)

###Propiedades de la matriz densidad

Una matriz se puede considerar como un operador densidad (rho) siempre que cumpla las siguientes propiedades:

i- sea hermitica, $\rho^{\dagger}$ = $\rho$.

ii- posea traza unitaria, Tr($\rho$) = 1.

iii- sea positiva definida, es decir, sus valores propios son no negativos.

iv- Si además representa un estado puro debe ser idempotente, es decir, $\rho^{2}=\rho$ [2]. 

####[2] L. de la Peña, M. Villavicencio, "Problemas y ejercicios de Mecánica cuántica", Fondo de cultura económica, México, Secc. XXI, pág. 643, 2003.  

Aplicando las propiedades a la matriz $\rho$ anterior

####i-

In [16]:
adj(rho)

2x2 Array{Intervalo,2}:
 (8.999999999999999999999999999999999999999999999999999999999999999999999999999793e-02, 6.400000000000000000000000000000000000000000000000000000000000000000000000000083e-01)    …  (-3.200000000000000000000000000000000000000000000000000000000000000000000000000041e-01, -2.999999999999999999999999999999999999999999999999999999999999999999999999999931e-02)
 (-3.200000000000000000000000000000000000000000000000000000000000000000000000000041e-01, -2.999999999999999999999999999999999999999999999999999999999999999999999999999931e-02)     (9.999999999999999999999999999999999999999999999999999999999999999999999999999725e-03, 1.600000000000000000000000000000000000000000000000000000000000000000000000000021e-01)  

####ii-

In [17]:
Tr(rho)

(9.999999999999999999999999999999999999999999999999999999999999999999999999999698e-02, 8.000000000000000000000000000000000000000000000000000000000000000000000000000104e-01)

####iv-

In [18]:
rho^2

2x2 Array{Intervalo,2}:
 (9.999999999999999999999999999999999999999999999999999999999999999999999999999455e-04, 1.28000000000000000000000000000000000000000000000000000000000000000000000000006e-01)  …  (2.999999999999999999999999999999999999999999999999999999999999999999999999999836e-03, 2.56000000000000000000000000000000000000000000000000000000000000000000000000012e-01) 
 (2.999999999999999999999999999999999999999999999999999999999999999999999999999836e-03, 2.56000000000000000000000000000000000000000000000000000000000000000000000000012e-01)     (8.999999999999999999999999999999999999999999999999999999999999999999999999999577e-03, 5.120000000000000000000000000000000000000000000000000000000000000000000000000239e-01)

Vemos que la $\rho$ construida no cumple con las propiedades descritas anteriormente. Por lo tanto, $\rho$ no es un operador densidad.

##Sistemas de ecuaciones lineales con intervalos

Consideremos un sistema de ecuaciones lineales general de la forma

\begin{eqnarray}
a_{11}x_{1} + a_{12}x_{2} &=& b_{1},\\
a_{21}x_{1} + a_{22}x_{2} &=& b_{2}.
\end{eqnarray}

Eliminando $x_{1}$ mediante un procedimiento análogo al de la eliminación gaussiana, podemos resolver para $x_{2}$ aplicando la ecuación

\begin{equation}
( a_{22}-(a_{21}/a_{11}) a_{12}	)x_{2} = b_{2}-(a_{21}/a_{11})b_{1}	
\end{equation}

De esta manera, utilizando el valor de $x_{2}$ podemos resolver para $x_{1}$ mediante la ecuación

\begin{equation}
a_{11}x_{1}=b_{1}-a_{12}x_{2}		
\end{equation}

Ejemplo: Consideremos el siguiente sistema de ecuaciones, 

\begin{eqnarray}
2.0000x_{1} + 3.001x_{2} &=& 1.0000,\\
0.6667x_{1} + 1.000x_{2} &=& 0.3333,
\end{eqnarray}

del cual nos proponemos determinar $x_{2}$. Eliminando $x_{1}$ en la segunda ecuación, obtenemos

\begin{equation}
( 1.000-(0.6667/2.0000)3.001)x_{2} = 0.3333 - (0.6667/2.0000)1.0000	
\end{equation}

Calculando para $n=4$ cifras significativas [3]:

####[3] R. E. Moore, R. B. Kearfott, M. J. Cloud, "Introduction to Interval Analysis", SIAM editorial, Philadelphia, Cap.7, Secc.7.2, pág. 90, 2009.

In [22]:
#calculando en primera instancia la división de la ecuación anterior
0.6667/2.000

0.33335

El valor exacto de la división en la ecuación anterior es $0.6667/2.0000 = 0.33335$. Sin embargo el objetivo aquí es usar intervalos, por lo que debemos escribir que
\begin{equation}
0.6667/2.0000 \in [0.3333, 0.3334]. 
\end{equation}

De nuevo, para $(0.6667/2.0000)\times 3.001$ obtenemos el valor exacto $1.00038335$, pero en términos de intervalos queda

\begin{equation}
(0.6667/2.0000)\times 3.001 \in [1.000, 1.001]. 
\end{equation}

Finalmente, los valores completos son

\begin{equation}
1.000-(0.6667/2.0000)\times 3.001 \in [-0.0010, 0], 
\end{equation}

\begin{equation}
0.3333-(0.6667/2.0000)\times 1.0000 \in [-0.0001, 0]. 
\end{equation}

Por consiguiente, $x_{2}$ esta dado por
\begin{equation}
x{2} \in [-0.0001, 0]/[-0.0010, 0] = [0, \infty].
\end{equation}

Para $n=5$ cifras significativas:

\begin{equation}
0.6667/2.0000 \in [0.33335, 0.33335], 
\end{equation}

\begin{equation}
(0.6667/2.0000)\times 3.001 \in [1.0003, 1.0004], 
\end{equation}

\begin{equation}
1.000-(0.6667/2.0000)\times 3.001 \in [-0.00005, -0.00005], 
\end{equation}

\begin{equation}
0.3333-(0.6667/2.0000)\times 1.0000 \in [-0.00040, -0.00030]. 
\end{equation}

\begin{equation}
\therefore x{2} \in [-0.00040, -0.00030]/[-0.00005, -0.00005] = [0.\underline{1}2500, 0.\underline{1}6667].
\end{equation}

In [6]:
#usando las operaciones sobre intervalos construidas en el modulo Intervalos, obtenemos efectivamente
Intervalo(-0.00005,-0.00005)/Intervalo(-0.00040,-0.00030)

(1.25e-01, 1.666666666666666666666666666666666666666666666666666666666666666666666666666674e-01)

el cual es el resultado obtenido para $n=5$ cifras significativas.

Luego,

Para $n=6$, $x_{2} \in [0.\underline{1}28205, 0.\underline{1}31579]$;

Para $n=7$, $x_{2} \in [0.\underline{130}2083, 0.\underline{130}5484]$;

Para $n=8$, $x_{2} \in [0.\underline{1304}1210, 0.\underline{1304}4613]$;

Para $n=9$, $x_{2} \in [0.\underline{13042911}1, 0.\underline{13042911}2]$.

Por lo tanto, vemos que al aumentar el número de cifras significativas disminuye el error del intervalo otorgando un valor de $x_{2}$ cada vez más acertado.